### ISSUE TITLE
Implement a `rollback` command

### ISSUE THREAD

**Issue by joeduffy** [pulumi/pulumi#96]
## Hello!

- Vote on this issue by adding a 👍 reaction
- To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already)

## Issue details

I am envisioning rollbacks to be as simple as reapplying changes from old snapshots/deployment records.  If you use Git for versioning of deployment records, then it is pretty straightforward.

For example, imagine we have tagged a stable release as `v1.2`:

    $ git tag v1.2
    $ pulumi update prod
    ... 1.2 is in production now ...

Now later on, we've done some things, and horked our environment:

    $ git commit
    $ pulumi update prod
    ... oops! ...

To get back to a sane state:

    $ git checkout v1.2
    $ pulumi update prod
    ... 1.2 is back in production, we're good ...

At the very least, we need to document the "playbook" for how to go about this.  And it's not quite as simple as shown above, because the rollback will need the old deployment record (which will have been obliterated by the `git checkout v1.2` -- unless we are using a Pulumi deployment server).  Perhaps this is left as-is, since it's plumbing and you need to know what you're doing.

To make this nicer, we might consider a rollback command which is just porcelain on top:

    $ pulumi deploy rollback prod v1.2

This effectively would do the same as above, but it automatically smuggles the current targets file so that you don't need to do anything manual.  Pretty cool demo too!

**Comment by joeduffy** [comment]
We frequently get asked by potential customers about whether Pulumi supports rollback. The short answer is "yes", since you can always accomplish rollback by rolling forward to a prior program plus configuration. The longer answer is "not what you probably meant" -- an opportunity to do better.

There are two primary scenarios in which I've heard people want rollback.

The first is the ability to trigger a rollback when something has gone wrong. Particularly with [a `history` command](https://github.com/pulumi/pulumi/issues/636), and the ability to [export specific checkpoints from it](https://github.com/pulumi/pulumi/issues/1910), it would seem we have some of the pieces needed to do this. This calls into question, however, whether it's valid to use a goal state from a checkpoint rather than running a program. It could very well be that our answer is "re-deploy the code," but we will need to understand whether this is an adequate answer (certainly when I've told some customers this -- especially those not using a Git-oriented workflow -- it was not).

The second is the ability to automatically rollback changes for a partially applied update. This would act more like AWS CloudFormation. Presumably, given our system, we'd need to rewind back to the goal state as it existed just prior to performing an update. And I am skeptical that we should do this by default, so it would probably materialize as a flag such as `--rollback-on-failure` or somesuch.

The two are related, and it would be wonderful if they leveraged similar mechanisms in how they work.

**Comment by thejosephstevens** [comment]
Hey there! My team is evaluating whether/when to move to Pulumi from TF, and one of the things I'm interested in that Terraform can't currently do is being able to rollback safely from a failed rollout. The reason I think it's compelling here is that the direct integration between Pulumi and a proper coding language makes it much more interesting to me to go ahead and write the rollback code myself as part of a custom provider. If as part of the provider definition you had the option of defining a callback function from each of your CRUD definitions that could get passed the little bit of state necessary to perform a rollback (helping with the less straightforward cases like undoing a delete), that would be excellent.

Longer term and more generally, I'm interested in options to get my hands more in the weeds of how Pulumi applies the actual state once it's been evaluated. In particular, things like complex destructive resource updates with a rollover migration in the middle (obviously out of scope of this particular issue)

**Comment by XBeg9** [comment]
@joeduffy would be great to see up/down hooks (as any ORMs do with their migrations). This way anyone can inject anything that must happen if a deployment fails for each particular Component.

**Comment by lukehoban** [comment]
> would be great to see up/down hooks (as any ORMs do with their migrations). This way anyone can inject anything that must happen if a deployment fails for each particular Component.

That is more or less what https://github.com/pulumi/pulumi/issues/1691 is tracking - does that look like it's in the ballpark of what you are looking for?  Would be great to add any details on requirements there.

**Comment by lukehoban** [comment]
My current proposal for how we'll enable this:
1. Add the ability to use a statefile as a target for an update (instead of a program) to the engine.  This may require capturing more metadata in the state file than is currently available to be fully accurate - including the items in https://github.com/pulumi/pulumi/issues/7285.  It will also rely on using provider configuration from the state file in a similar way to `destroy` and `refresh` which also do not run a program.
2. Add a `pulumi up --rollback-to <update-id>` or `pulumi rollback <update-id>` command to do an update which uses the support in (1) to drive toward the previous update id as the desired state instead of a output of a program.
3. Provide an option to `pulumi up` like `--rollback-on-failure` which automates triggering a `pulumi up --rollback-to <updated-id>` on the base update id of the `up` operation if the `up` fails.  Notably of course, this follow-up could itself still fail, and the stack would be left in that state.

**Comment by pgavlin** [comment]
> Add the ability to use a statefile as a target for an update (instead of a program) to the engine. This may require capturing more metadata in the state file than is currently available to be fully accurate - including the items in Consider storing ignoreChanges and other goal state options into state #7285. It will also rely on using provider configuration from the state file in a similar way to destroy and refresh which also do not run a program.

I'm worried that this approach will not work well for providers that require data flow between resources, especially in the case that rolling back requires replacements of resources whose outputs feed other resources' inputs. In those cases, it is extremely unlikely that using a state file as an update source would produce a useable result: it is likely that property values from replaced resources that flow into other resources will have changed with respect to the state file.

I do think that this approach could work well for e.g. k8s, where data flow between resources is much less common.

**Comment by Frassle** [comment]
As Pat said above given that resource inputs may derive from other resource outputs and that any rollback operation could change those outputs and given that Pulumi can't see those derivations (because they're part of the user program) I can't see how we could possibly do this in general.
I suppose we could give a program to go from a state file to the best effort pulumi program that would result in that state file which users could opt into using with the caveat that they need to look carefully to see how much information was lost in how those properties were set because in most cases things like ids are going to be wrong.

**Comment by dixler** [comment]
At least for pulumi maintainers, this would help with setting pulumi to a prior state so that we can run an update to get it to the current state and easily target a code path without having to re-edit the program.

**Comment by illegalnumbers** [comment]
Is this in progress? It would be helpful.

**Comment by ghostsquad** [comment]
Isn't a rollback just `git revert` followed by another `apply`?

**Comment by hansarh** [comment]
This would be really nice! We've had and incident where pulumi refresh was running on a branch which removed resources from state

**Comment by mbsulliv** [comment]
I came to this thread thinking that this is exactly the feature I needed, but the more I thought about it, isn't the whole point of "infrastructure as code" to have the code reflect the infrastructure?  If you roll back the infrastructure without also rolling back the code, won't you just undo your rollback the next time you run "pulumi update"?

My current thoughts on this are that I need to be diligent in checking in working configurations in source control so I can use those configurations to roll back changes as needed.

**Comment by nloum** [comment]
@mbsulliv I agree, but the only caveat is that you always have to be careful to add new providers , commit that and do a pulumi up, and wait to add new resources with said providers until a subsequent commit. This is because you can end up trying to delete resources that don't have any provider associated with them. That's what #15370 is about.

**Comment by Frassle** [comment]
I've raised https://github.com/pulumi/docs/issues/16036 to document that the way to do this is via source control. There's not enough information in pulumi to be able to rollback to previous state in general so implementing a `rollback` command isn't something we can do.
